nrf-hal-common/src/uarte.rs

```diff
@@ -406,18 +432,249 @@ pub enum Error {
+
+        // Prevent writing to buffer while DMA transfer is in progress.
+        if uarte.events_txstarted.read().bits() == 1 {
+            return Err(nb::Error::WouldBlock);
+        }
+
+        if self.written < self.tx_buf.len() {
+            self.tx_buf[self.written] = b;
+            self.written += 1;
+            Ok(())
+        } else {
+            Err(nb::Error::WouldBlock)
+        }
+    }
+
+    /// Flush the TX buffer non-blocking. Returns nb::Error::WouldBlock if not yet flushed.
+    fn flush(&mut self) -> nb::Result<(), Self::Error> {
+        let uarte = unsafe { &*T::ptr() };
+
+        // If txstarted is set, we are in the process of transmitting.
+        let in_progress = uarte.events_txstarted.read().bits() == 1;
+
+        if in_progress {
+            let endtx = uarte.events_endtx.read().bits() != 0;
+            let txstopped = uarte.events_txstopped.read().bits() != 0;
+            if endtx || txstopped {
+                // We are done, cleanup the state.
+                uarte.events_txstarted.reset();
+                self.written = 0;
+
+                // Conservative compiler fence to prevent optimizations that do not
+                // take in to account actions by DMA. The fence has been placed here,
+                // after all possible DMA actions have completed.
+                compiler_fence(SeqCst);
+
+                if txstopped {
+                    return Err(nb::Error::Other(Error::Transmit));
+                }
+
+                // Lower power consumption by disabling the transmitter once we're
+                // finished.
+                stop_write(uarte);
+
+                compiler_fence(SeqCst);
+                Ok(())
+            } else {
+                // Still not done, don't block.
+                Err(nb::Error::WouldBlock)
+            }
+        } else {
+            // No need to trigger transmit if we don't have anything written
+            if self.written == 0 {
+                return Ok(());
+            }
+
+            start_write(uarte, &self.tx_buf[0..self.written]);
+
+            Err(nb::Error::WouldBlock)
+        }
+    }
+}
+
+// Auto-implement the blocking variant
+impl<T> bserial::write::Default<u8> for UarteTx<T> where T: Instance {}
+
+impl<T> core::fmt::Write for UarteTx<T>
+where
+    T: Instance,
+{
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        s.as_bytes()
+            .iter()
+            .try_for_each(|c| nb::block!(self.write(*c)))
+            .map_err(|_| core::fmt::Error)
+    }
+}
+
+impl<T> serial::Read<u8> for UarteRx<T>
+where
+    T: Instance,
+{
+    type Error = Error;
+    fn read(&mut self) -> nb::Result<u8, Self::Error> {
+        let uarte = unsafe { &*T::ptr() };
+
+        compiler_fence(SeqCst);
+
+        let in_progress = uarte.events_rxstarted.read().bits() == 1;
+        if in_progress && uarte.events_endrx.read().bits() == 0 {
+            return Err(nb::Error::WouldBlock);
```

Comment:
If we wait for ENDRX, we'll end up returning `WouldBlock` even if data has been received, right? That seems like an issue, especially since there is no timeout logic.

Reply:
Yes, that is a consequence of the embedded hal interface which doesn't allow you to read more than one byte at a time, so using rx_buf with size > 1 doesn't make sense right now. I was thinking to make this a bit more advanced where it would fill the rx_buf if it was empty, otherwise read from it until there are no more bytes to read. I'll push a fix for that.

Reply:
Hmm, without some sort of timer, the RX would not be able read more than 1 byte at a time, as it seems it won't ever set  ENDRX until the expected number of bytes is received. I didn't originally want to require a timer for this, so I wonder if we should instead restrict to reading 1 byte no matter how big the rx_buf is, at least for now?

Reply:
Yeah, I think we'll have to make it read one byte at a time.

Reply:
Ok, done. Added a comment about the reason for that.